### ISSUE TITLE
[downgrade] steps without names are given invalid identifiers and no name

### ISSUE THREAD

**Issue by jimsheldon** [drone/go-convert#124]
Here is an example v1 pipeline
```yaml
name: test
stages:
- name: test
  spec:
    clone: {}
    platform:
      arch: amd64
      os: linux
    runtime:
      spec: {}
      type: cloud
    steps:
    - spec:
        run: pip3 install
      type: script
    - spec:
        run: echo hello
      type: script
  type: ci
version: 1
```
Which downgrades to this
```yaml
pipeline:
  identifier: default
  name: default
  orgIdentifier: default
  projectIdentifier: default
  properties:
    ci:
      codebase:
        build: <+input>
  stages:
  - stage:
      identifier: test
      name: test
      spec:
        cloneCodebase: true
        execution:
          steps:
          - step:
              spec:
                command: pip3 install
              timeout: ""
              type: Run
          - step:
              identifier: "1"
              spec:
                command: echo hello
              timeout: ""
              type: Run
        platform:
          arch: Amd64
          os: Linux
        runtime:
          spec: {}
          type: Cloud
      type: CI
```

The first step is missing both `name` and `identifier` fields which are required.

For the second step, the `name` field is missing, and `"1"` is not a valid identifier.

The pipeline YAML editor shows this error for the `identifier`:
```
String does not match the pattern of "^[a-zA-Z_][0-9a-zA-Z_]{0,127}$".
```

It should downgrade to something like this
```yaml
pipeline:
  identifier: default
  name: default
  orgIdentifier: default
  projectIdentifier: default
  properties:
    ci:
      codebase:
        build: <+input>
  stages:
  - stage:
      identifier: test
      name: test
      spec:
        cloneCodebase: true
        execution:
          steps:
          - step:
              identifier: "step1"
              name: "step 1"
              spec:
                command: pip3 install
              timeout: ""
              type: Run
          - step:
              identifier: "step2"
              name: "step 2"
              spec:
                command: echo hello
              timeout: ""
              type: Run
        platform:
          arch: Amd64
          os: Linux
        runtime:
          spec: {}
          type: Cloud
      type: CI
```

In other CI solutions like github and circle, they will show the first X characters of the command. Often, when users intentionally omit the "name" field, the commands are short and this works well.

Perhaps we could do something like this? Ensuring that `identifier` satisfies the above regex pattern.

```yaml
pipeline:
  identifier: default
  name: default
  orgIdentifier: default
  projectIdentifier: default
  properties:
    ci:
      codebase:
        build: <+input>
  stages:
  - stage:
      identifier: test
      name: test
      spec:
        cloneCodebase: true
        execution:
          steps:
          - step:
              identifier: pip3install
              name: pip3install
              spec:
                command: pip3 install
              timeout: ""
              type: Run
          - step:
              identifier: echohello
              name: echo hello
              spec:
                command: echo hello
              timeout: ""
              type: Run
        platform:
          arch: Amd64
          os: Linux
        runtime:
          spec: {}
          type: Cloud
      type: CI
```

**Comment by bradrydzewski** [comment]
Here is how we handle this for stages:

```
		ID: d.identifiers.Generate(
			slug.Create(stage.Id),
			slug.Create(stage.Name),
			slug.Create(stage.Type),
		),
```

Note that we pass in the Id, Name and Type.  If the Id is empty, it will attempt to generate the identifier from the Name.  If the Name is empty it will attempt to generate the identifier from the Type.

So we should be able to fix here:
https://github.com/drone/go-convert/blob/master/convert/harness/downgrader/downgrade.go#L443C1-L444C25

```diff
	spec_ := src.Spec.(*v1.StepExec)
-	var id = d.identifiers.Generate(
-		slug.Create(src.Name))
	return &v0.Step{
-		ID:      id,
+		ID: d.identifiers.Generate(
+			slug.Create(src.Id),
+			slug.Create(src.Name),
+			slug.Create(src.Type),
+		),
```

@eoinmcafee00 can you audit the code to follow the above pattern?  I can see some other areas that might also be subject to this issue:
* <https://github.com/drone/go-convert/blob/master/convert/harness/downgrader/downgrade.go#L443>
* <https://github.com/drone/go-convert/blob/master/convert/harness/downgrader/downgrade.go#L476>
* <https://github.com/drone/go-convert/blob/master/convert/harness/downgrader/downgrade.go#L538>
* <https://github.com/drone/go-convert/blob/master/convert/harness/downgrader/downgrade.go#L559>

Note that the `identifiers.Generate` will automatically handle the scenario where duplicate identifiers exist, so the below code is not necessary and can be simpliefied

```
-	var id string
-	var name string
-
-	if src.Name != "" {
-		id = d.identifiers.Generate(slug.Create(src.Name))
-		name = convertName(src.Name)
-	} else {
-		id = d.identifiers.Generate(fmt.Sprintf("step%d", index))
-		name = fmt.Sprintf("step%d", index)
-	}

+	id := d.identifiers.Generate(
+		slug.Create(src.Id),
+		slug.Create(src.Name),
+		slug.Create(src.Type),
+	)

+	name := src.Name
+	if src.Name = "" {
+		src.Name = id
+	}

	return &v0.Step{
		ID:      id,
		ID:      name,
```

this should also eliminate the need to pass index:

```diff
-func (d *Downgrader) convertStepPlugin(src *v1.Step, stageEnv map[string]string, index int) *v0.Step {
+func (d *Downgrader) convertStepPlugin(src *v1.Step, stageEnv map[string]string) *v0.Step {
```

**Comment by jimsheldon** [comment]
Fix in #131
